pkg/test/framework/components/ingress/kube.go

```diff
@@ -42,113 +45,221 @@ var (
+	}, retryTimeout, retryDelay)
+	if err != nil {
+		return ""
+	}
+	return address.(string)
 }
 
-func (c *kubeComponent) Call(path string) (CallResponse, error) {
-	client := &http.Client{
-		Timeout: 1 * time.Minute,
+// HTTPSAddress returns HTTPS address of ingress gateway.
+func (c *kubeComponent) HTTPSAddress() string {
+	address, err := retry.Do(func() (interface{}, bool, error) {
+		return getHTTPSAddressInner(c.env, c.namespace)
+	}, retryTimeout, retryDelay)
+	if err != nil {
+		return ""
 	}
+	return address.(string)
+}
 
-	if !strings.HasPrefix(path, "/") {
-		path = "/" + path
+// createClient creates a client which sends HTTP requests or HTTPS requests, depending on
+// ingress type. If host is not empty, the client will resolve domain name and verify server
+// cert using the host name.
+func (c *kubeComponent) createClient(options CallOptions) (*http.Client, error) {
+	client := &http.Client{
+		Timeout: options.Timeout,
 	}
-	url := c.address + path
+	if options.CallType != PlainText {
+		scopes.Framework.Debug("Prepare root cert for client")
+		roots := x509.NewCertPool()
+		ok := roots.AppendCertsFromPEM([]byte(options.CaCert))
+		if !ok {
+			return nil, fmt.Errorf("failed to parse root certificate")
+		}
+		tlsConfig := &tls.Config{
+			RootCAs:    roots,
+			ServerName: options.Host,
+		}
+		if options.CallType == Mtls {
+			cer, err := tls.X509KeyPair([]byte(options.Cert), []byte(options.PrivateKey))
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse private key and server cert")
+			}
+			tlsConfig.Certificates = []tls.Certificate{cer}
+		}
+		tr := &http.Transport{
+			TLSClientConfig: tlsConfig,
+			DialTLS: func(netw, addr string) (net.Conn, error) {
+				if addr == options.Host+":443" {
+					addr = options.Address + ":443"
+				}
+				tc, err := tls.Dial(netw, addr, tlsConfig)
+				if err != nil {
+					scopes.Framework.Errorf("TLS dial fail: %v", err)
+					return nil, err
+				}
+				if err := tc.Handshake(); err != nil {
+					scopes.Framework.Errorf("SSL handshake fail: %v", err)
+					return nil, err
+				}
+				return tc, nil
+			}}
+		client.Transport = tr
+	}
+	return client, nil
+}
 
-	scopes.Framework.Debugf("Sending call to ingress at: %s", url)
+// createRequest returns a request for client to send, or nil and error if request is failed to generate.
+func (c *kubeComponent) createRequest(options CallOptions) (*http.Request, error) {
+	url := options.Address + options.Path
+	if options.CallType != PlainText {
+		url = "https://" + options.Host + ":443" + options.Path
+	}
 
 	req, err := http.NewRequest("GET", url, nil)
 	if err != nil {
+		return nil, err
+	}
+	if options.Host != "" {
+		req.Host = options.Host
+	}
+
+	scopes.Framework.Debugf("Created a request to send %v", req)
+	return req, nil
+}
+
+// sanitizeCallOptions checks and fills fields in CallOptions.
+func sanitizeCallOptions(options *CallOptions) {
+	if options.Timeout <= 0 {
+		options.Timeout = DefaultRequestTimeout
```

Comment:
Nit: give this responsibility to CallOptions
```
func(o *CallOptions) sanitize() error {
...
}

Reply:
Nice catch. Thanks!